### PR TITLE
[JENKINS-73499] Add a warning if there is a risk of exposing credentials through a non-TLS proxy connection

### PR DESCRIPTION
**Pull request by jmdesprez**
See [JENKINS-73499](https://issues.jenkins.io/browse/JENKINS-73499).

### Testing done

Here is a screenshot of my local testing:
image

For now there is no automated testing. Let me know if I need to add one (I will need some guidance).

### Proposed changelog entries

- Add a warning if there is a risk of exposing credentials through a non-TLS proxy connection

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] ~New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.~
- [ ] ~New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).~
- [ ] ~For dependency updates, there are links to external changelogs and, if possible, full differentials.~
- [ ] ~For new APIs and extension points, there is a link to at least one consumer.~
```

### Desired reviewers

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

**Review by PereBueno — APPROVED** [review]

File: core/src/main/java/hudson/ProxyConfiguration.java
Comment (PereBueno) [inline core/src/main/java/hudson/ProxyConfiguration.java:562]: Might be redundant, maybe enough with showing this validation in the password field?

File: core/src/main/java/hudson/ProxyConfiguration.java
Comment (Wadeck) [inline core/src/main/java/hudson/ProxyConfiguration.java:550]: 💡 I would recommend you to avoid naming method "doXxx" if they are not intended to be used as Stapler endpoint.

File: core/src/main/resources/hudson/Messages_fr.properties
Comment (Wadeck) [inline core/src/main/resources/hudson/Messages_fr.properties:112]: Are you sure the single quote is OK here? The other values in the file have two single quotes.
(I don't remember the rules for the encoding there, just a warning ;-) )

File: core/src/main/resources/hudson/Messages.properties
Comment (Wadeck) [inline core/src/main/resources/hudson/Messages.properties:113]: 💡 Is "be exposed" sufficient in terms of information? I mean, as someone configuring it, do I understand easily "where"/"to who" it is exposed?

**Review by Wadeck — COMMENTED** [review]
Just some non-blocker comments :-)

**Review by mawinter69 — COMMENTED** [review]
The warning is always shown as soon as username or password are not empty. Even when I configure a https proxy url or (as in the screenshot) no url is configured.
Ideally this warning is only shown when using a http url.

File: core/src/main/resources/hudson/Messages_fr.properties
Comment (jmdesprez) [reply core/src/main/resources/hudson/Messages_fr.properties]: Yes, you are right, sorry. Fixed in 203e3b2ebbaf6adfdf6abb3173c3c2a1a8d327e5

**Comment by jmdesprez** [comment]
> The warning is always shown as soon as username or password are not empty. Even when I configure a https proxy url 

My understanding is that it is not possible to configure an `https` proxy. When filling out the configuration, one only provides the host and the port number, like in the below screenshot:
image

And it is not possible to specify which protocol the server will use:
image

So I've come to the conclusion that the connection to the proxy is always via `http` (non-TLS). 

But I may have missed something.

> or (as in the screenshot) no url is configured. Ideally this warning is only shown when using a http url.

My example is a bit silly because If there is no URL configured, then the entire proxy configuration will be discarded anyway. 
Typically, the first two fields are filled in first, then the username and password. 

I can, of course, change this, but then I suppose the validation will be more complex to write because several fields will be involved.

File: core/src/main/resources/hudson/Messages.properties
Comment (jmdesprez) [reply core/src/main/resources/hudson/Messages.properties:113]: I tried to keep it short, but I can try to make it more clear. How about this one?

```
Please note that the connection between Jenkins and the proxy is always via an HTTP connection. Therefore, this insecure connection may expose the credentials provided.
```

Comment (Wadeck) [reply core/src/main/resources/hudson/Messages.properties:113]: As it's a warning, the conciseness is not "that" important I would say. The proposal works better for me :)

Comment (jmdesprez) [reply core/src/main/resources/hudson/Messages.properties]: @Wadeck I updated the messages: e0cf56459ca6c8d0d732358e4de27955d5d7858e

File: core/src/main/resources/hudson/Messages.properties
Comment (timja) [inline core/src/main/resources/hudson/Messages.properties:113]: maybe something like this? we shouldn't be including `please note` its verbose for no added value, the validation messages should be more direct:

```suggestion
ProxyConfiguration.NonTLSWarning=Jenkins only supports using an http connection to the proxy. The credentials may be exposed to someone on the same network.
```

**Review by timja — COMMENTED** [review]
I don't know how common https proxies are in the wild but maybe support for that should be added to go with this?

I've seen this:
https://www.chromium.org/developers/design-documents/secure-web-proxy/

and

https://security.stackexchange.com/a/23584

File: core/src/main/resources/hudson/Messages.properties
Comment (jmdesprez) [reply core/src/main/resources/hudson/Messages.properties]: Thanks for the suggestion, I updated the messages in b1780f7628c2d16c2eb2082787a803affb8b8c20.

**Review by timja — APPROVED** [review]

**Comment by jmdesprez** [comment]
> I don't know how common https proxies are in the wild but maybe support for that should be added to go with this?
> 
> I've seen this: [chromium.org/developers/design-documents/secure-web-proxy](https://www.chromium.org/developers/design-documents/secure-web-proxy/)
> 
> and
> 
> [security.stackexchange.com/a/23584](https://security.stackexchange.com/a/23584)

Based on (non-exhaustive) search on the Internet, https proxies are indeed not so common. On the client side, in addition to those you have mentioned, `curl` supports https proxies (see [daniel.haxx.se/blog/2016/11/26/https-proxy-with-curl](https://daniel.haxx.se/blog/2016/11/26/https-proxy-with-curl/)). For example, OKHttp doesn't support it (see [square/okhttp#3787](https://github.com/square/okhttp/issues/3787)).

> maybe support for that should be added to go with this?

Do you mean as part of this ticket or in a new ticket?

**Comment by jmdesprez** [comment]
Additionally, it seems that people usually don't worry about the http connection because the proxy is usually on a private network.

**Comment by jmdesprez** [comment]
@timja 
I resolved the conversations and I updated the branch. 
Please let me know if anything is missing so it can be marked as ready for merge.

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
